File: discover.go
package main

import (
	"os"
	"path/filepath"
)

var composeFiles = []string{
	"docker-compose.yml",
	"docker-compose.yaml",
	"compose.yml",
	"compose.yaml",
}

func discoverServices(config *Config) []string {
	exe, err := os.Executable()
	if err != nil {
		return nil
	}
	scanDir := filepath.Dir(exe)
	parentDir := filepath.Dir(scanDir)
	selfDir := filepath.Base(scanDir)

	entries, err := os.ReadDir(parentDir)
	if err != nil {
		return nil
	}

	var services []string
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		if entry.Name() == selfDir {
			continue
		}

		fullPath := filepath.Join(parentDir, entry.Name())
		for _, cf := range composeFiles {
			composePath := filepath.Join(fullPath, cf)
			if _, err := os.Stat(composePath); err == nil {
				services = append(services, composePath)
				break
			}
		}
	}

	return services
}

File: src/discover.ts
import { readdirSync, statSync } from "node:fs";
import { join, basename } from "node:path";
import type { Config } from "./types.js";

const COMPOSE_FILES = [
  "docker-compose.yml",
  "docker-compose.yaml",
  "compose.yml",
  "compose.yaml",
];

export function discoverServices(config: Config): string[] {
  const entries = readdirSync(config.scanPath);
  const services: string[] = [];

  for (const entry of entries) {
    const fullPath = join(config.scanPath, entry);
    if (!statSync(fullPath).isDirectory()) continue;
    if (config.skipDirs.includes(entry)) continue;

    for (const composeFile of COMPOSE_FILES) {
      const composePath = join(fullPath, composeFile);
      try {
        statSync(composePath);
        services.push(composePath);
        break;
      } catch {}
    }
  }

  return services;
}
